From 1449bcdffb6b318e488e12553428af48c96547a5 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:04:52 -0400
Subject: [usb] usbtest: Fix BOS control test for USB 2.01 devices

Message-id: <1410887463-20674-221-git-send-email-dzickus@redhat.com>
Patchwork-id: 95305
O-Subject: [RHEL7 PATCH 220/591] usbtest: Fix BOS control test for USB 2.01 devices.
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 8cf4328569acc37ac5c5b4eb27ae86c3758f627b
Author: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Date:   Fri Dec 13 13:44:17 2013 -0800

    usbtest: Fix BOS control test for USB 2.01 devices.

    Commit c952a8ba7136505cd1ca01735cc748ddc08c7d2f "usb: usbtest: add a
    test case to support bos for queue control" will cause USB 2.01 and USB
    2.10 devices with a BOS descriptor to fail case 15 of the control test.

    The Link PM errata (released in 2007, updated in 2011) says:

    "The value of the bcdUSB field in the standard USB 2.0 Device Descriptor
    is used to indicate that the device supports the request to read the BOS
    Descriptor (i.e.  GetDescriptor(BOS)). Devices that support the BOS
    descriptor must have a bcdUSB value of 0201H or larger."

    The current code says that non-SuperSpeed devices *must* return -EPIPE,
    as this comment shows:

                    /* sign of this variable means:
                     *  -: tested code must return this (negative) error code
                     *  +: tested code may return this (negative too) error code
                     */
                    int                     expected = 0;

    This means the test will fail with USB 2.01 and USB 2.10 devices that
    provide a BOS descriptor.  Change it to only require a stall response if
    the USB device bcdUSB is less than 2.01.

    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
    Acked-by: Huang Rui <ray.huang@amd.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/misc/usbtest.c b/drivers/usb/misc/usbtest.c
index bff058e..446ff55 100644
--- a/drivers/usb/misc/usbtest.c
+++ b/drivers/usb/misc/usbtest.c
@@ -1224,7 +1224,7 @@ test_ctrl_queue(struct usbtest_dev *dev, struct usbtest_param *param)
     len = le16_to_cpu(udev->bos->desc->wTotalLength);
    else
     len = sizeof(struct usb_bos_descriptor);
-   if (udev->speed != USB_SPEED_SUPER)
+   if (le16_to_cpu(udev->descriptor.bcdUSB) < 0x0201)
     expected = -EPIPE;
    break;
   default:
-- 
1.7.1